src/GovUk.Education.ExploreEducationStatistics.Data.Model/Migrations/20240802123234_Routine_RebuildIndexes.sql

```diff
@@ -0,0 +1,283 @@
+      FROM (SELECT ips.avg_fragmentation_in_percent AS FragPercent
+            FROM #TableList AS tl
+                   JOIN sys.dm_db_index_physical_stats(DB_ID(), NULL, NULL, NULL, 'LIMITED') ips
+                        ON ips.object_id = OBJECT_ID(tl.ObjectName, 'U')
+                   JOIN sys.indexes idx ON idx.object_id = ips.object_id AND idx.index_id = ips.index_id
+                   JOIN sys.objects obj ON obj.object_id = ips.object_id
+                   JOIN sys.schemas sch ON sch.schema_id = obj.schema_id
+            WHERE ips.alloc_unit_type_desc = 'IN_ROW_DATA'
+              AND tl.ObjectName = @ObjectName
+              AND sch.Name = @SchemaName
+              AND idx.name = @IndexName) AS Fragmented
+      WHERE __RebuildIndexesAlterIndexes.RunId = @RunId
+        AND __RebuildIndexesAlterIndexes.Id = @StatsRowIndex;
+
+      -- Reorganise or rebuild
+      RAISERROR ('Executing command ''%s''.', 0, 1, @Command) WITH NOWAIT;
+      EXEC (@Command);
+
+      -- Save end info for reporting
+      UPDATE __RebuildIndexesAlterIndexes
+      SET __RebuildIndexesAlterIndexes.EndTime        = GETUTCDATE(),
+          __RebuildIndexesAlterIndexes.EndFragPercent = Fragmented.FragPercent,
+          __RebuildIndexesAlterIndexes.Info           = IIF(@FragPercentBefore >= @FragmentationThresholdRebuild,
+                                                            'Rebuilt', 'Reorganized')
+      FROM (SELECT ips.avg_fragmentation_in_percent AS FragPercent
+            FROM #TableList AS tl
+                   JOIN sys.dm_db_index_physical_stats(DB_ID(), NULL, NULL, NULL, 'LIMITED') ips
+                        ON ips.object_id = OBJECT_ID(tl.ObjectName, 'U')
+                   JOIN sys.indexes idx ON idx.object_id = ips.object_id AND idx.index_id = ips.index_id
+                   JOIN sys.objects obj ON obj.object_id = ips.object_id
+                   JOIN sys.schemas sch ON sch.schema_id = obj.schema_id
+            WHERE ips.alloc_unit_type_desc = 'IN_ROW_DATA'
+              AND tl.ObjectName = @ObjectName
+              AND sch.Name = @SchemaName
+              AND idx.name = @IndexName) AS Fragmented
+      WHERE __RebuildIndexesAlterIndexes.RunId = @RunId
+        AND __RebuildIndexesAlterIndexes.Id = @StatsRowIndex;
+
+      IF @StopTime < GETUTCDATE()
+        BEGIN
+          -- Update Info for all unprocessed indexes, because we're skipping them now we've hit @StopTime
+          UPDATE __RebuildIndexesAlterIndexes
+          SET Info = 'Skipped: stored procedure run time exceeded timeout'
+          WHERE Info IS NULL;
+
+          SET @SkipRemainingTasks = 1;
+        END
+
+      IF @StatsRowIndex >= @StatsCount
+        RAISERROR ('Completed optimising indexes for tables.', 0, 1, NULL) WITH NOWAIT;
+
+      SET @StatsRowIndex += 1;
+    END
+
+  /*-----------------------------------------------------------------------*/
+
+  SET @TablesRowIndex = 1;
+  SELECT @TablesCount = COUNT(Id) FROM #TableList;
```

Comment:
I think you might have missed a point that we only want to do `UPDATE STATISTICS` for tables that have had indexes touched by the first stage, but this refactoring to base this from `#TableList` instead of `#Stats` has changed this to do `UPDATE STATISTICS` for each table regardless.

I think this is unnecessary so I would revert this change unless you found something that says we should always be doing `UPDATE STATISTICS`?

Quite often on the run of this stored procedure, no indexes are touched because all the percentages fall below `FragmentationThresholdReorganize`.

If we wanted to do this every time regardless because there's no dependency on `#Stats` we could have looked two separate stored procedures would have probably made more sense.

Reply:
> I think you might have missed a point that we only want to do UPDATE STATISTICS for tables that have had indexes touched by the first stage

You're right I didn't realise this.

Rather than revert this entirely, I've added a check that skips the update stats query if no indexes were rebuilt/reorganized for the current table.